coalib/misc/Caching.py

```diff
@@ -0,0 +1,193 @@
+    with open(filename, "wb") as f:
+        return pickle.dump(data, f)
+
+
+def add_to_changed_files(log_printer, project_dir, changed_files):
+    """
+    Save the set of files in ``changed_files`` to disk for future use
+    in ``update_last_coala_run_time``.
+
+    :param log_printer:   A LogPrinter object to use for logging.
+    :param project_dir:   The root directory of the project on which
+                          coala is to be run.
+    :param changed_files: A set of files that had changed since the last
+                          time coala was run.
+    """
+    data = pickle_load(log_printer, changed_files_db, {})
+
+    if project_dir in data:
+        data[project_dir].update(changed_files)
+    else:
+        data[project_dir] = changed_files
+
+    pickle_dump(log_printer, changed_files_db, data)
+
+
+def get_last_coala_run_time(log_printer, project_dir):
+    """
+    Get the last time coala was run on the project in epoch format
+    (number of seconds since Jan 1, 1970).
+
+    :param log_printer: A LogPrinter object to use for logging.
+    :param project_dir: The root directory of the project on which
+                        coala is to be run.
+    :return:            Returns a dict of files with file path as key
+                        and the last time that coala was run on that file
+                        as value. Returns None if this is the first time coala
+                        is run on that project.
+    """
+    data = pickle_load(log_printer, caching_db, {})
+    return data.get(project_dir, None)
+
+
+def add_new_files_since_last_run(log_printer, project_dir, last_coala_run,
+                                 new_files):
+    """
+    Start tracking new files given in ``new_files`` by adding them to the
+    database.
+
+    :param log_printer:    A LogPrinter object to use for logging.
+    :param project_dir:    The root directory of the project on which
+                           coala is to be run.
+    :param last_run_coala: Previously available data to which the new files
+                           need to be appended.
+    :param new_files:      The list of new files that need to be tracked.
+                           These files are initialized with their last
+                           modified tag as -1.
+    """
+    if last_coala_run is None:
+        last_coala_run = {}
+    for new_file in new_files:
+        last_coala_run[new_file] = -1
+    data = pickle_load(log_printer, caching_db, {})
+    data[project_dir] = last_coala_run
+    pickle_dump(log_printer, caching_db, data)
+
+
+def update_last_coala_run_time(log_printer, project_dir):
+    """
+    Update the last time coala was run on the project for each file
+    to the current time.
+
+    :param log_printer: A LogPrinter object to use for logging.
+    :param project_dir: The root directory of the project on which
+                        coala is to be run.
+    """
+    current_time = calendar.timegm(time.gmtime())
```

Comment:
How would we do this? Keep a separate record of when coala was run last time? Then we could check the current date with the last run date and purge the cache if last run is greater than the current time?